docs/repo/crates/discv4.md

```diff
@@ -0,0 +1,354 @@
+
+[File: bin/reth/src/node/mod.rs]()
+```rust ignore
+  pub async fn execute(&self) -> eyre::Result<()> {
+    //--snip--
+    let network = config
+        .network_config(db.clone(), chain_id, genesis_hash, self.network.disable_discovery)
+        .start_network()
+        .await?;
+
+    info!(peer_id = ?network.peer_id(), local_addr = %network.local_addr(), "Started p2p networking");
+
+    //--snip--
+    }
+```
+
+During this process, a new `NetworkManager` is created through the `NetworkManager::new()` function, which starts the discovery protocol through a handful of newly spawned tasks. Lets take a look at how this actually works under the hood. 
+
+[File: ]()
+```rust ignore
+impl<C> NetworkManager<C>
+where
+    C: BlockProvider,
+{
+    //--snip--
+
+    pub async fn new(config: NetworkConfig<C>) -> Result<Self, NetworkError> {
+        let NetworkConfig {
+            //--snip--
+            secret_key,
+            mut discovery_v4_config,
+            discovery_addr,
+            boot_nodes,
+            //--snip--
+            ..
+        } = config;
+
+        //--snip--
+
+        discovery_v4_config = discovery_v4_config.map(|mut disc_config| {
+            // merge configured boot nodes
+            disc_config.bootstrap_nodes.extend(boot_nodes.clone());
+            disc_config.add_eip868_pair("eth", status.forkid);
+            disc_config
+        });
+
+        let discovery = Discovery::new(discovery_addr, secret_key, discovery_v4_config).await?;
+
+    //--snip--
+    }
+}
+```
+
+First, the `NetworkConfig` is deconstructed and the `disc_config` is updated to merge configured [bootstrap nodes]() and add the `forkid` to adhere to [EIP 868](https://eips.ethereum.org/EIPS/eip-868). This updated configuration variable is then passed into the `Discovery::new()` function.
+
+[File: ]()
+```rust ignore
+impl Discovery {
+    /// Spawns the discovery service.
+    ///
+    /// This will spawn the [`reth_discv4::Discv4Service`] onto a new task and establish a listener
+    /// channel to receive all discovered nodes.
+    pub async fn new(
+        discovery_addr: SocketAddr,
+        sk: SecretKey,
+        discv4_config: Option<Discv4Config>,
+    ) -> Result<Self, NetworkError> {
+        let local_enr = NodeRecord::from_secret_key(discovery_addr, &sk);
+        
+        let (discv4, discv4_updates, _discv4_service) = if let Some(disc_config) = discv4_config {
+            let (discv4, mut discv4_service) =
+                Discv4::bind(discovery_addr, local_enr, sk, disc_config)
+                    .await
+                    .map_err(NetworkError::Discovery)?;
+            let discv4_updates = discv4_service.update_stream();
+            // spawn the service
+            let _discv4_service = discv4_service.spawn();
+            (Some(discv4), Some(discv4_updates), Some(_discv4_service))
+        } else {
+            (None, None, None)
+        };
```

Comment:
this will change a bit. I think it's sufficient to keep this high level

Reply:
Sounds good, I can update the chapter with this later today. 

Do you think it would be best to simply add placeholders for code snippets and remove the current snippets completely? 

Or do you think some of the snippets that are currently included should stay?

Reply:
atm this is very detailed and describes a lot of the code. I think, if we just cover this more broadly, like:

* there's the Service task that drives the udp socket and implements the Discv4 protocol and streams discovered nodes back to the NetworkManager, which delegates them to the peerset etc

it is sufficiently thorough. 

The service documentation is excellent, I'd even include this as rust docs directly.